### PR TITLE
NativeAOT SuperPMI collections

### PR DESCRIPTION
**Pull request by TIHan**
This adds NativeAOT SuperPMI collections runs for the `SmokeTests`. Will work on windows+linux x64+arm64.

### CODE REVIEW

File: src/coreclr/scripts/superpmi.py [hunk 1]
```diff
@@ -1101,6 +1129,141 @@ async def run_crossgen2(print_prefix, assembly, self):
...
+
+                    root_output_filename = make_safe_filename("nativeaot_" + nativeaot_output_filename + "_")
+
+                    with open(rsp_filepath, "a") as rsp_write_handle:
+                        rsp_write_handle.write(input_filepath + "\n")
+                        if self.coreclr_args.host_os.lower() == "windows":
+                            rsp_write_handle.write("--directpinvokelist:" + os.path.join(self.core_root, "build", "WindowsAPIs.txt") + "\n")
+                        rsp_write_handle.write("-o:" + nativeaot_output_assembly_filename + "\n")
+                        rsp_write_handle.write("-r:" + os.path.join(self.coreclr_args.nativeaot_aotsdk_path, "System.*.dll") + "\n")
+                        rsp_write_handle.write("-r:" + os.path.join(self.core_root, "System.*.dll") + "\n")
```
Comment (MichalStrehovsky) [inline src/coreclr/scripts/superpmi.py:1195]: Why is this rewriting needed?

Depending on what you need to do, the `--make-repro-path` ILC option might be useful. When `--make-repro-path:some\path\to\file.zip` is specified on the ILC command line (can be added by adding `<ItemGroup><IlcArg Include="--make-repro-path:some\path\to\file.zip" /></ItemGroup>` to the csproj, for example), the compiler will generate a ZIP file that has:

* All the input files
* An RSP file that can be directly passed to the compiler and contains relative paths to the all the input files.

The advantage of this zip is that one doesn't need to munge the RSP anymore. If you need to replace files, you could just replace things in the ZIP (e.g. you could go over all the DLLs in CORE_ROOT, find a file with a matching name in the ZIP and replace that) - the key point is that you wouldn't need to parse the RSP.

Comment (TIHan) [reply src/coreclr/scripts/superpmi.py:1195]: The rewriting was needed as the paths are completely different since this gets run on the helix machines. 

As an example, the rsp has a ref:
```
C:\work\runtime\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net8.0\Microsoft.CSharp.dll
```

but it needs to point to:
```
c:\work\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\Microsoft.CSharp.dll
```

> All the input files

Is this including the references?

Comment (TIHan) [reply src/coreclr/scripts/superpmi.py:1198]: Spent a few days looking at this.

There are things with `--make-repro-path` that make this hard:

- `--make-repro-path` only specifies the output directory, rather than the actual `.zip` name.
- The actual file name it produces starts with a HashCode + AssemblyName, so I don't know for sure what the output will be. I guess I could check for `EndsWith($"{AssemblyName}".zip)`, but the point is I can't predict what the actual `.zip` path will be.
- The `repro.rsp` that gets created and placed in the `.zip` also includes `--make-repro-path` which is something we do not want.

If we assume that I can predictably get the `.zip` file, I think zeroing-out the assemblies to be fake dummy files and then replacing them with matching ones from `Core_Root` or `aotsdk` folders is going to be much harder to handle than just simply munging the original `.ilc.rsp` files and cleaning up "-r", "-o", etc. It's only a handful lines of code, instead of having code in superpmi-collect-setup.py and superpmi.py to manage zeroing-out and replacing files.

I do understand the want to avoid parsing the `.ilc.rsp` files as it does create extra maintenance if you all change the `.ilc.rsp` which *could* break the collections. I still think this is the better option for the time being.

An alternative would be that we create our own `.rsp` files, which is what we do with crossgen2 collections. The issue with this approach is that we do not know if a ".dll" has a `Main` method and it's managed without an extra tool to tell us otherwise. We would also miss out on an extra knobs and other command arguments that may be impactful, but that's the case with crossgen2 as well.

Comment (MichalStrehovsky) [reply src/coreclr/scripts/superpmi.py:1198]: > The repro.rsp that gets created and placed in the .zip also includes --make-repro-path which is something we do not want.

I think this was a bug that got fixed in #90935.

> --make-repro-path only specifies the output directory, rather than the actual .zip name.

This could be addressed by always emitting this into a new directory you control - there will be a single file in it once ILC finishes and it will have what you need.

> I do understand the want to avoid parsing the .ilc.rsp files as it does create extra maintenance if you all change the .ilc.rsp which could break the collections. I still think this is the better option for the time being.

As long as the JIT team will handle all of the breaks that can happen here, I'm fine with whatever option. The only reason I proposed `--make-repro-path` is that it will likely be the most stable interface. I can imagine things getting more complex than `-r` replacement once we go out of nativeaot/SmokeTests to the rest of the test tree where we have tests that consist of multiple assemblies, etc.

> We would also miss out on an extra knobs and other command arguments that may be impactful, but that's the case with crossgen2 as well.

I think using the rsp file generated by the SDK is most representative to customer scenarios. I wouldn't want to replicate the crossgen2 testing approach here.

Comment (TIHan) [reply src/coreclr/scripts/superpmi.py:1198]: > As long as the JIT team will handle all of the breaks that can happen here, I'm fine with whatever option.

Yes, the JIT team will always be responsible for how these collections are handled - I see no expectations from other teams having to deal with it.

> I think this was a bug that got fixed in https://github.com/dotnet/runtime/pull/90935.

Very cool!

> I think using the rsp file generated by the SDK is most representative to customer scenarios. I wouldn't want to replicate the crossgen2 testing approach here.

I agree. I would rather us use the generated RSP file for this reason. At some point, we should also do the same thing to crossgen2.

We can improve this incrementally. The goal now is just to get something simple that works and has value.

**Comment by TIHan** [comment]
@dotnet/jit-contrib @BruceForstall this is ready.

File: eng/pipelines/coreclr/templates/run-superpmi-collect-job.yml [hunk 2]
```diff
@@ -100,6 +100,12 @@ jobs:
     - ${{ if eq(parameters.collectionName, 'libraries_tests') }}:
       - name: InputDirectory
         value: '$(Build.SourcesDirectory)/artifacts/tests/libraries/$(osGroup).$(archType).$(buildConfigUpper)'
+    - ${{ if and(eq(parameters.collectionName, 'smoke_tests'), eq(parameters.osGroup, 'windows')) }}:
+      - name: InputDirectory
+        value: '$(Build.SourcesDirectory)\artifacts\tests\coreclr\obj\${{ parameters.osGroup }}.${{ parameters.archType }}.${{ parameters.buildConfig }}\Managed\nativeaot\SmokeTests'
+    - ${{ if and(eq(parameters.collectionName, 'smoke_tests'), ne(parameters.osGroup, 'windows')) }}:
+      - name: InputDirectory
+        value: '$(Build.SourcesDirectory)/artifacts/tests/coreclr/obj/${{ parameters.osGroup }}.${{ parameters.archType }}.$(buildConfigUpper)/Managed/nativeaot/SmokeTests'
```
Comment (BruceForstall) [inline eng/pipelines/coreclr/templates/run-superpmi-collect-job.yml:108]: The `libraries_tests` case above doesn't require windows and non-windows cases. So I would expect this doesn't, either.
```suggestion
    - ${{ if eq(parameters.collectionName, 'smoke_tests')) }}:
      - name: InputDirectory
        value: '$(Build.SourcesDirectory)/artifacts/tests/coreclr/obj/$(osGroup).$(archType).$(buildConfigUpper)/Managed/nativeaot/SmokeTests'
```

File: src/coreclr/inc/jiteeversionguid.h [hunk 3]
```diff
@@ -43,11 +43,11 @@ typedef const GUID *LPCGUID;
...
-constexpr GUID JITEEVersionIdentifier = { /* a2974440-e8ee-4d95-9e6e-799a330be1a0 */
-    0xa2974440,
-    0xe8ee,
-    0x4d95,
-    {0x9e, 0x6e, 0x79, 0x9a, 0x33, 0x0b, 0xe1, 0xa0}
+constexpr GUID JITEEVersionIdentifier = { /* 43ca1203-3a26-4d8a-89ec-60a3654bb2ad */
+    0x43ca1203,
+    0x3a26,
+    0x4d8a,
+    {0x89, 0xec, 0x60, 0xa3, 0x65, 0x4b, 0xb2, 0xad}
```
Comment (BruceForstall) [inline src/coreclr/inc/jiteeversionguid.h:50]: You can revert the change to this file.

File: src/coreclr/scripts/superpmi.py [hunk 4]
```diff
@@ -1101,6 +1129,144 @@ async def run_crossgen2(print_prefix, assembly, self):
...
 
+            ################################################################################################ Do collection using nativeaot
+            if self.coreclr_args.nativeaot is True:
+                logging.debug("Starting collection using nativeaot")
+
+                async def run_nativeaot(print_prefix, original_rsp_filepath, self):
+                    if not original_rsp_filepath.endswith(".ilc.rsp"):
+                        raise RuntimeError(f"Expected a '.ilc.rsp' file for nativeaot, but got {original_rsp_filepath}")
+
+                    rsp_filepath = os.path.join(self.temp_location, make_safe_filename("nativeaot_" + original_rsp_filepath) + ".rsp")
```
Comment (BruceForstall) [inline src/coreclr/scripts/superpmi.py:1140]: It would be useful to have some comments here describing what is going to happen with the code that follows: the expected contents of the .rsp file, what transformations are required and why, etc.

**Review by BruceForstall — COMMENTED** [review]
Generally LGTM

**Comment by jakobbotsch** [comment]
@TIHan Can you please check the superpmi-collect runs you have been triggered for errors? E.g. I see all the "libraries pmi" runs failing with
```
Console log: '0' from job c74f489f-b165-4186-a8d9-e8f2f0f058df workitem 43264b0b-93c7-4f6b-b133-873825bb3b3b (ubuntu.2204.amd64.rt) executed on machine a0001QC running Linux-5.15.0-1041-azure-x86_64-with-glibc2.35
chmod: cannot access '/datadisks/disk1/work/AC490993/w/A220090C/e/*.sh': No such file or directory
+ /usr/bin/python3 /datadisks/disk1/work/AC490993/p/superpmi/superpmi.py collect --clean -log_level DEBUG --pmi -pmi_location /datadisks/disk1/work/AC490993/p/superpmi/pmi.dll -pmi_path /datadisks/disk1/work/AC490993/p/superpmi/R2RDump /datadisks/disk1/work/AC490993/p/superpmi/R2RTest /datadisks/disk1/work/AC490993/p/superpmi/crossgen2 /datadisks/disk1/work/AC490993/p/superpmi/xunit -assemblies /datadisks/disk1/work/AC490993/w/A220090C/u/binaries -arch x64 -build_type checked -core_root /datadisks/disk1/work/AC490993/p/superpmi -output_mch_path /datadisks/disk1/work/AC490993/w/A220090C/uploads/libraries.pmi.0.linux.x64.checked.mch -log_file /datadisks/disk1/work/AC490993/w/A220090C/uploads/libraries.pmi.0.linux.x64.checked.log
[00:29:45] ================ Logging to /datadisks/disk1/work/AC490993/w/A220090C/uploads/libraries.pmi.0.linux.x64.checked.log
Specify `-ilc_rsps` if `--nativeaot` is given
+ export _commandExitCode=1
+ exit 1
['0' END OF WORK ITEM LOG: Command exited with 1]
```

([link](https://helixri107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-heads-will-nativeaot-collc74f489fb1654186a8/0/1/console.962a1e43.log?sv=2021-08-06&se=2023-12-06T00%3A17%3A16Z&sr=c&sp=rl&sig=3SsAeYjxW3b0ZeLWyGeHCTrq%2B9E7iMK%2FcBhCR%2BCLO%2Bg%3D))

**Comment by BruceForstall** [comment]
I looked at a recent collection run and see a few concerning things:

1.
```
2023-09-07T00:19:46.5692728Z cp: cannot create regular file '/mnt/vss/_work/1/s/artifacts/obj/': Not a directory
2023-09-07T00:19:46.5692958Z cp: cannot create regular file '/mnt/vss/_work/1/s/artifacts/obj/': Not a directory
2023-09-07T00:19:46.5693150Z /mnt/vss/_work/1/s/eng/native/version/copy_version_files.sh: line 29: /mnt/vss/_work/1/s/artifacts/obj/_version.c: No such file or directory
```

Why are these failing? What script/process is the invoker?

2. Failures during collection
```
Unhandled exception: System.IO.DirectoryNotFoundException: Could not find a part of the path '/mnt/vss/_work/1/s/artifacts/tests/coreclr/obj/linux.x64.Checked/Managed/nativeaot/SmokeTests/SharedLibrary/SharedLibrary/native/SharedLibrary.exports'.
...
Unhandled exception: System.IO.DirectoryNotFoundException: Could not find a part of the path '/mnt/vss/_work/1/s/src/tests/nativeaot/SmokeTests/DynamicGenerics/rd.xml'.
...
Unhandled exception: System.CommandLine.CommandLineException: '/mnt/vss/_work/1/s/src/tests/nativeaot/SmokeTests/TrimmingBehaviors/NonEmbedded.ILLink.Descriptor.xml' doesn't exist
```

Are these files built when the smoke tests are built? Are they just not getting copied to Helix?

**Comment by TIHan** [comment]
> Why are these failing? What script/process is the invoker?

This one I am not sure.

> Are these files built when the smoke tests are built?

Except for the `.exports' file, the `.xml` files are located where the source of the tests are. Which means we would have to do more work to get these files copied over.

> Are they just not getting copied to Helix?

I updated it to at least copy over the `.exports` file, but I can't copy the `.xml` ones because they are not part of the output of the tests.

Seeing how there are multiple tests, when built, that will not include the non-DLL files they depend on in the output, it may make sense to do something similar to what @MichalStrehovsky suggested. I was hoping that it would be just the assemblies and the pinvokelist.

@BruceForstall What do you recommend we do? We get tens of thousands of method contexts so it still works well enough - we can iterate on it over time.

**Review by BruceForstall — APPROVED** [review]
LGTM
